source/extensions/certificate_providers/certificate_provider.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace CertificateProviders {
+
+struct Certpair {
+  const std::string& certificate_;
+  const std::string& private_key_;
+};
+
+struct Capabilites {
+  /* whether or not a provider provides a ca cert directly */
+  bool provide_ca_cert = false;
+
+  /* Whether or not a provider provides identity certs directly */
+  bool provide_identity_certs = false;
+
+  /* whether or not a provider supports generating identity certs during handshake,
+   * which requires the capability provide_ca_cert to sign the certificates
+   */
+  bool generate_identity_certs = false;
+};
+
+class CertificateProvider {
+public:
+  virtual ~CertificateProvider() = default;
+
+  virtual Capabilites capabilities() const PURE;
+
+  /**
+   * @return CA certificate from provider
+   */
+  virtual const std::string& getCACertificate() const PURE;
```

Comment:
I think both of these methods need to take a parameter, which is the name of the certificate to obtain.  The caller would use this to pass the value of the [`CertificateProviderPluginInstance.certificate_name` field](https://github.com/envoyproxy/envoy/blob/7900682f09b25163445e83042fe7068ba0bbbd32/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L269).

Note that not all certificate provider impls will actually use this field; many providers (such as the static provider) will support only one certificate per instance, which means that the cert name passed in is irrelevant.  However, there may be other providers that support multiple certs and need the name to differentiate between them.  This may be particularly important in the case of an ingress proxy that supports many virtual domains, each with their own certs; in that case, it's desirable to configure the cert name with each VirtualHost from the control plane instead of having to configure a separate cert provider in the bootstrap file for each cert.

Reply:
got it